feat(tag): add language field and findByLanguageUuid method (#481)

AngelPelaezS1 opened the pull request:
User Story: #481

Summary
This PR adds a new idLanguage field to the TagDocument and TagDto classes to support language tagging. Additionally, it introduces a new repository method findByIdLanguage(UUID idLanguage) to query tags by their associated language UUID.

Changes

TagDocument.java
Added a new field idLanguage of type UUID, mapped to the language field in the MongoDB document.

TagDto.java

Added a new field idLanguage of type UUID, serialized as uuid_language in JSON.

TagRepository.java

Added method findByIdLanguage(UUID idLanguage) to query tags by language.

TagRepositoryTest.java

Added unit tests to verify the correct behavior of findByIdLanguage.

Unit tests

findByIdLanguage is covered by unit tests to ensure expected repository behavior. No integration tests added as this change is limited to the repository and DTO/document layers.

Note on unrelated changes

The changes made to the logging in DatabaseUpdater.java were removed, as they were not related to the main goal of adding the language field to TagDocument. These changes were not necessary for the PR and have been discarded.

Next Steps

The following PRs will address updates to the service and controller layers to fully integrate the language field into the tag management functionality.

Commit:
deleted unnecessary method addChallengeToSolved from UserServiceImpl

Commit:
fixed wrong index JsonProperty

Commit:
adapted ChallengeSolvedController for auto-call

Commit:
updated ChallengeServiceImpl to don't need userId

Commit:
Modified the method to increase everytime is called

Commit:
creation interface IChallengeService

Commit:
creation service ChallengeServiceImpl

Commit:
added new exceptions

Commit:
creation test for ChallengeServiceImpl

Commit:
added library okhttp3 for testing

Commit:
added challenge.service.url

RustyGearBox committed:
modified save solution method

FerGimenezRoglia committed:
Add Fernando to contributors.md

# Conflicts:
#	contributors.md

Commit:
fix order buildAndExpand

Commit:
fixed tests

Commit:
reversed last fix

Commit:
reversed last fix

Commit:
update changelog

JofreCoca committed:
Update contributors.md

Commit:
updated ChallengeSolvedController

IngaD89 committed:
Detect updated microservices

AngelPelaezSanz committed:
Feature/481: Add language field to Tag and findByLanguageUuid repository method

AngelPelaezSanz committed:
Feature/481: Add language field to Tag and findByLanguageUuid repository method

ivilarop committed:
refactor: adapt ChallengeSolvedController to updated timesSolved logic and document in changelog

AngelPelaezSanz committed:
Feature/481: Add language field to Tag and findByLanguageUuid repository method

AngelPelaezSanz committed:
Feature/481: Add language field to Tag and findByLanguageUuid repository method

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

IngaD89 commented:
**Note:** SonarCloud is currently failing — please check the tests and the database connection.

# Review of the `TagDocument` model:

In general terms, the implementation is consistent and aligned with the rest of the project code.

## Points to consider:

- In the database, the field `language` is stored as:  
````
@Field(name = "language") 
private UUID idLanguage;
````
This is not an error, but for semantic clarity (especially since it’s a UUID), it is recommended to consider a more descriptive name like `language_id`.

- Are tags always associated with a language or could there be standalone tags?  
If every tag must be linked to a language, it's a good idea to mark the field with @NotNull to prevent creating orphan tags.

# Review of the `TagDto`:

This is not a functional error, but it can cause semantic confusion and mapping errors if naming consistency is not maintained.  
In the DTO:  
````
@JsonProperty("uuid_language")
private UUID idLanguage;
````
In the model:  
````
@Field(name = "language")
private UUID idLanguage;
````

Suggestion: unify the field name (for example, `language_id`) both in the model and the DTO to improve clarity and avoid issues during serialization/deserialization.

# Business Logic Questions:

Given the user story description:  
“Currently, tags are generic. They must be per language”  
And considering this task is part of the epic related to challenge creation and visualization, I have the following questions:  

- Can a tag be associated with multiple languages?  
  Example: The tag `"backend"` could be related to both Java and PHP.  
  If yes, the current model (`TagDocument` with a single `idLanguage`) does not support this, and a many-to-many relationship would need to be represented.

- Can a language have multiple associated tags?  
  Example: Java could have the tags `"backend"`, `"OOP"`, `"Spring"`...  
  If that's the case, this is well covered by performing a reverse lookup on TagDocument using the idLanguage field to find all associated tags.

- Should tags be assigned both to challenges and languages?  
  Currently, the `ChallengeDocument` model has a list of tag IDs. This would allow associating any tag to any challenge regardless of language.  
  If tags should match the languages selected in the challenge, this coherence should be validated in the business logic (for example, in the service that creates the challenge).

Before continuing with the full review, I need to clarify these questions to validate whether the model design aligns with the business logic.

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

AngelPelaezSanz committed:
Feature/481: updated all test constructors to include language UUID in TagDocument instances

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

AngelPelaezS1 commented:
Thanks for the detailed review! I’ve addressed the following points:

Renamed the field in TagDocument from language to language_id using @Field(name = "language_id") for semantic clarity.

Updated TagDto to use @JsonProperty("language_id") to ensure consistency with the model and avoid mapping issues.

Confirmed that a tag can only be associated with one language, so the current model is valid.

Regarding the validation to ensure that tags used in a challenge match one of the challenge’s languages, I’ll implement that when I work on the ChallengeService.

Let me know if anything’s missing or unclear — thanks again!

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

AngelPelaezSanz committed:
Feature/481: refactor: rename language field to language_id in TagDocument and TagDto

IngaD89 commented:
I see the field was renamed to language_id and the DTO is consistent. It’s also clear that each tag is linked to only one language, so the model looks good.

Just a quick note: when I tested the getAllTags method (I wanted to make sure the new field is retrieved correctly), I ran into a runtime error about UUID conversion from MongoDB — specifically, a ConverterNotFoundException trying to convert from org.bson.types.Binary to java.util.UUID.
This might be due to how UUIDs are stored or handled in the database, and it could cause issues later on.
You might not have reached this part of the development yet, but it's something to keep in mind as you continue working on the feature.

Nice work, keep going!

ivilarop committed:
Merge pull request #885 from IT-Academy-BCN/feature/415-Modify-method-to-increase-times-solved-counter

feat (challenge): Modify method to increase times solved counter (#415)

FerGimenezRoglia committed:
Add Fernando Giménez to contributors.md

ivilarop committed:
delete ChallengeNotFoundException

ivilarop committed:
Merge pull request #894 from IT-Academy-BCN/feature/fernando-onboarding

chore(contributors): add Fernando Giménez Roglia to contributors.md (#446)

IngaD89 committed:
update microservice names in config and add Java/Gradle setup steps

ivilarop committed:
Merge pull request #890 from IT-Academy-BCN/feature/413-create-methods-to-comunicate-user.micro-and-challenge.micro

feat (user): Create methods to comunicate user.micro and challenge.micro (#413)

ivilarop committed:
Merge pull request #897 from IT-Academy-BCN/chore(contributors)-add-jofrecoca-to-contributors.md

chore(contributors): add Jofre Coca contributors.md(#454)

ivilarop committed:
Merge pull request #893 from IT-Academy-BCN/feature/417-modifing-the-method-of-saving-solution-to-connect-to-micro.challenge

feat (user): Modifing the method of saving solution to connect to micro.challenge (#417)

ivilarop committed:
Merge pull request #901 from IT-Academy-BCN/feature/464-detect-microservice-conf

feat(workflow): detect updated microservices in CI (#464)

ivilarop commented:
The PR title does not follow the project naming conventions.
Please update it to follow the format:

type(module): short description (#taskNumber)

👀 This PR seems to be part of a larger task, focused on updating the model and repository layer.
If this is a split PR, consider updating the title to reflect that (e.g. PR 1/2 - feat(tag): add language field and query method (#481)), so it's clearer for reviewers and for future traceability.

Also, you could briefly mention in the description what will be addressed in the following PRs (e.g., service or controller updates) if they’re coming.

ivilarop commented:
📌 This PR adds a new field (idLanguage) to the Tag model and a repository method to filter tags by language UUID.
As it introduces a behavioral change in how tags are queried, consider adding a changelog entry

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/config/dbchangelog/DatabaseUpdater.java:
```diff
@@ -106,9 +106,13 @@ public void addFieldToAllDocuments(ReactiveMongoTemplate reactiveMongoTemplate)
         ).doOnSuccess(result -> {
-            logger.info("Matched count: {}", result.getMatchedCount());
-            logger.info("Modified count: {}", result.getModifiedCount());
-        }).doOnError(error -> logger.error(ERROR_UPDATE, error.getMessage())).subscribe();
```
⚠️ The changes to DatabaseUpdater.java improve logging, but appear unrelated to the primary goal of this logging request (adding the language field to TagDocument).

If it's a code improvement, it shouldn't be addressed and should be audited as technical debt.

Are these changes necessary for the purpose of the PR?

AngelPelaezS1 replied:
I made these logging improvements based on a recommendation from SonarQube to enhance traceability and code monitoring. However, I understand that for the primary goal of this PR, which is to add the language field to TagDocument, they are not necessary. Therefore, I can revert these changes if they are not relevant to the purpose of the PR.

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/config/dbchangelog/DatabaseUpdater.java:
```diff
@@ -117,8 +121,12 @@ public void removeFieldToAllDocuments(ReactiveMongoTemplate reactiveMongoTemplat
                 .doOnSuccess(result -> {
-                    logger.info("Matched count: {}", result.getMatchedCount());
-                    logger.info("Modified count: {}", result.getModifiedCount());
+                    if (result != null && result.wasAcknowledged()) {
```
⚠️ The changes to DatabaseUpdater.java improve logging, but appear unrelated to the primary goal of this logging request (adding the language field to TagDocument).

If it's a code improvement, it shouldn't be addressed and should be audited as technical debt.

Are these changes necessary for the purpose of the PR?

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/repository/TagRepository.java:
```diff
@@ -3,13 +3,13 @@
 @Repository
 public interface TagRepository extends ReactiveMongoRepository<TagDocument, UUID> {
-
     Mono<TagDocument> findByTagName(String tagName);
```
Does this method apply to the changes you are implementing?

AngelPelaezS1 replied:
I didn't add that method, and it doesn't apply to the changes I'm implementing.

AngelPelaezSanz committed:
Feature/481: implement: remove unrelated changes in DatabaseUpdater.java and update changelog

AngelPelaezSanz committed:
Merge remote-tracking branch 'origin/feature/481-add-lenguage' into feature/481-add-language-to-tag

AngelPelaezS1 commented:
Close
